src/acl.c

```diff
@@ -1438,23 +1681,37 @@ int ACLCheckPubsubPerm(const user *u, robj **argv, int idx, int count, int liter
     /* If we survived all the above checks, the user can execute the
      * command. */
     return ACL_OK;
-
 }
 
 /* Check whether the command is ready to be executed by ACLCheckCommandPerm.
  * If check passes, then check whether pub/sub channels of the command is
  * ready to be executed by ACLCheckPubsubPerm */
-int ACLCheckAllUserCommandPerm(const user *u, struct redisCommand *cmd, robj **argv, int argc, int *idxptr) {
-    int acl_retval = ACLCheckCommandPerm(u,cmd,argv,argc,idxptr);
-    if (acl_retval != ACL_OK)
-        return acl_retval;
-    if (cmd->proc == publishCommand)
-        acl_retval = ACLCheckPubsubPerm(u,argv,1,1,0,idxptr);
-    else if (cmd->proc == subscribeCommand)
-        acl_retval = ACLCheckPubsubPerm(u,argv,1,argc-1,0,idxptr);
-    else if (cmd->proc == psubscribeCommand)
-        acl_retval = ACLCheckPubsubPerm(u,argv,1,argc-1,1,idxptr);
-    return acl_retval;
+int ACLCheckAllUserCommandPerm(user *u, struct redisCommand *cmd, robj **argv, int argc, int *idxptr) {
+    listIter li;
+    listNode *ln;
+    if (u == NULL) return ACL_OK;
+
+    int closest_error = ACL_DENIED_CMD;
+
+    /* Check all of the selectors */
+    listRewind(u->selectors,&li);
+    while((ln = listNext(&li))) {
+        aclSelector *s = (aclSelector *) listNodeValue(ln);
+        int acl_retval = ACLSelectorCheckCmd(s,cmd,argv,argc,idxptr);
+        if (acl_retval == ACL_OK) {
+            if (cmd->proc == publishCommand)
+                acl_retval = ACLSelectorCheckPubsub(s,argv,1,1,0,idxptr);
+            else if (cmd->proc == subscribeCommand)
+                acl_retval = ACLSelectorCheckPubsub(s,argv,1,argc-1,0,idxptr);
+            else if (cmd->proc == psubscribeCommand)
+                acl_retval = ACLSelectorCheckPubsub(s,argv,1,argc-1,1,idxptr);
+            if (acl_retval == ACL_OK) {
+                return ACL_OK;
+            }
+        }
+        if (acl_retval > closest_error) closest_error = acl_retval;
```

Comment:
"closest" means "most detailed"?
maybe add some comment explaining what we aim for...
p.s. don't we also need to keep the correct `idxptr` of the error we choose to propagate upwards?
(i suppose the same command can't do both keys and channels, so i guess it's just a theoretical problem, but it stands out when looking at the coed)

Reply:
In either case, I defined the logic to be the first "key/channel" that was not matched by any selector. That should also handle the case if modules want to define a command that touches both.